api/logic/evm_test.go

```diff
@@ -0,0 +1,93 @@
+package logic
+
+import (
+	_ "embed"
+	"encoding/hex"
+	"encoding/json"
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	ethCommon "github.com/ethereum/go-ethereum/common"
+	"github.com/stretchr/testify/require"
+
+	"github.com/oasisprotocol/oasis-indexer/analyzer/evmabi"
+)
+
+//go:embed test_contracts/artifacts/Varied.json
+var artifactVariedJSON []byte
+var Varied *abi.ABI
+
+func init() {
+	type artifact struct {
+		ABI *abi.ABI
+	}
+	var artifactVaried artifact
+	if err := json.Unmarshal(artifactVariedJSON, &artifactVaried); err != nil {
+		panic(err)
+	}
+	Varied = artifactVaried.ABI
+}
+
+func TestEVMParseTypes(t *testing.T) {
+	data, err := hex.DecodeString("a85acffaffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000000000001ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000010101010101010101010101010101010101010101010101010101010101010101000000000000000000000000010101010101010101010101010101010101010101010101010101010101010101010101010101010202020200000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000001c00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000024000000000000000000000000000000000000000000000000000000000000002800000000000000000000000000000000000000000000000000000000000000001010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000016100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000016100000000000000000000000000000000000000000000000000000000000000")
+	require.NoError(t, err)
+	method, args, err := EVMParseData(data, Varied)
+	require.Equal(t, Varied.Methods["test"], *method)
+	jsonExpected := []string{
+		"-1",     // int8
+		"1",      // uint8
+		"\"-1\"", // int256
+		"\"1\"",  // uint256
+		"true",   // bool
+		"\"AQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQE=\"", // bytes32
```

Comment:
You mentioned somewhere else: "possibly weird philosophically: uint8[] and uint8[n] I believe will also be base64"

I think the philosophy as we discussed it is pretty straightforward: This JSON formatting is primarily intended to look good to a human reader, and we don't really expect it to be re-parsed often.
In that vein, I'd say that we should render any array/sequence of bytes (types `bytes`, `bytes1..bytes32`, `uint8[]` (maybe this is the same as `bytes`?), `uint8[N]`) as a hex string. I don't think eth devs ever expect to see their byte blobs rendered as base64.

Reply:
displaying byte string as hex is for the frontend :upside_down_face:. this is the separate problem of getting a little document over some format that fits in an HTTP response, and we've selected JSON for that. JSON doesn't support byte strings which this document contains, and for whatever reason, the very conventional Go and/or oasis-core thing to do is to use base64 for that :skull:

Reply:
actually do we already use hex in most of our byte-strings-in-json for nexus?

Reply:
> displaying byte string as hex is for the frontend

But if frontend wanted to do that, it would need to traverse the whole JSON object, and either a) understand and follow the ABI, replicating much of the work you're doing here (in which case, why even bother here?), or b) heuristically decide which values are likely easiest undertood when displayed as hex but were relayed in base64 (ugly). It's not a lot of work to re-encode here, right? A few extra cases in `evmPreMarshal()`?

The main reason we're constructing this JSON is so frontend (any frontend, not just ours), or humans in general, can make better sense of the values. So let's take it all the way?

This is perhaps inconsistent with the base64 examples I posted above ... But I'd argue that those are more about offering a uniformly-, predictably-encoded machine-readable representation of the events, rather than producing eye candy. Also, speaking of consistency, I see this PR already hex-encodes `address` types (courtesy of `go-ethereum`).

Reply:
message from 3 weeks in the future: lmao fine